system/trace/trace.c

```diff
@@ -0,0 +1,775 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <sys/ioctl.h>
+#include <nuttx/lib/regex.h>
+#include <nuttx/note/notectl_driver.h>
+#ifdef CONFIG_DRIVER_NOTERAM
+#  include <nuttx/note/noteram_driver.h>
+#endif
+
+#include "trace.h"
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: notectl_enable
+ ****************************************************************************/
+
+static bool notectl_enable(int flag, int notectlfd)
+{
+  struct note_filter_mode_s mode;
+  int oldflag;
+
+  ioctl(notectlfd, NOTECTL_GETMODE, (unsigned long)&mode);
+
+  oldflag = (mode.flag & NOTE_FILTER_MODE_FLAG_ENABLE) != 0;
+  if (flag == oldflag)
+    {
+      /* Already set */
+
+      return false;
+    }
+
+  if (flag)
+    {
+      mode.flag |= NOTE_FILTER_MODE_FLAG_ENABLE;
+    }
+  else
+    {
+      mode.flag &= ~NOTE_FILTER_MODE_FLAG_ENABLE;
+    }
+
+  ioctl(notectlfd, NOTECTL_SETMODE, (unsigned long)&mode);
+
+  return true;
+}
+
+/****************************************************************************
+ * Name: trace_cmd_start
+ ****************************************************************************/
+
+static int trace_cmd_start(int index, int argc, FAR char **argv,
+                           int notectlfd)
+{
+  FAR char *endptr;
+  int duration = 0;
+  bool cont = false;
+
+  /* Usage: trace start [-c][<duration>] */
+
+  if (index < argc)
+    {
+      if (strcmp(argv[index], "-c") == 0)
+        {
+          cont = true;
+          index++;
+        }
+    }
+
+  if (index < argc)
+    {
+      duration = strtoul(argv[index], &endptr, 0);
+      if (!duration || endptr == argv[index] || *endptr != '\0')
+        {
+          fprintf(stderr,
+                  "trace start: invalid argument '%s'\n", argv[index]);
+          return ERROR;
+        }
+
+      index++;
+    }
+
+  /* Clear the trace buffer */
+
+#ifdef CONFIG_DRIVER_NOTERAM
```

Comment:
how about we change trace.h like this:
```
#ifdef CONFIG_DRIVER_NOTERAM
void trace_dump_clear(void);
#else
#define trace_dump_clear
#endif
```
so we don't need check CONFIG_DRIVER_NOTERAM again and again. other functions in trace.h could apply the approach.